### ISSUE TITLE
inline tail call does not honor comptime semantics

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#12108]
Zig Version: `0.10.0-dev.2992+1fee9eac8`

```
test "inline tail call" {
    if (!foo()) {
        @compileError("expected a comptime value");
    }
}

inline fn foo() bool {
    return bar();
}

inline fn bar() bool {
    return true;
}
```

Should pass, however, the result of `foo()` ends up being runtime-known:

```
$ stage2/bin/zig test test3.zig 
test3.zig:10:9: error: expected a comptime value
        @compileError("expected a comptime value");
        ^
```

Problem disappears if `foo` is rewritten like this:

```zig
inline fn foo() bool {
    const result = bar();
    return result;
}
```

**Comment by nektro** [comment]
doesnt pass on stage1 either. is this a new feature of `inline fn`s being eagerly evaluated?

**Comment by andrewrk** [comment]
Yes it's new `inline` semantics that are going into the language specification. Function inlining happens during semantic analysis, meaning that if an inlined function returns a comptime-known value, it is comptime-known at the callsite.

**Comment by tauoverpi** [comment]
With the new behaviour, would `inline fn assert(ok: bool) void` make more sense than the current implementation given that the condition may be comptime known and thus all code following it can be deleted (replaced by panic or a return in release-fast/small)?

**Comment by mlugg** [comment]
Appears to be fixed in master.
